src/cascadia/Remoting/Monarch.cpp

```diff
@@ -518,20 +563,42 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
                                   TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
                                   TraceLoggingKeyword(TIL_KEYWORD_TRACE));
 
-                return mruWindowArgs.PeasantID();
+                result = mruWindowArgs.PeasantID();
+                break;
             }
-            positionInList++;
         }
 
-        // Here, we've checked all the windows, and none of them was both alive
-        // and the most recent (on this desktop). Just return 0 - the caller
-        // will use this to create a new window.
-        TraceLoggingWrite(g_hRemotingProvider,
-                          "Monarch_getMostRecentPeasantID_NotFound",
-                          TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
-                          TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+        lock.unlock();
+
+        if (peasantsToErase.size() > 0)
+        {
+            std::unique_lock uniqueLock{ _mruPeasantsMutex };
+
+            // I wish I had std::erase_if (C++20) but I don't yet
+            // partition the array into [not dead | dead ]
+            auto partition = std::stable_partition(_mruPeasants.begin(), _mruPeasants.end(), [&](const auto& p) {
+                const auto id = p.PeasantID();
+                const auto it = std::find(peasantsToErase.cbegin(), peasantsToErase.cend(), id);
+                // keep the element if it was not found in the list to erase.
+                return it == peasantsToErase.cend();
+            });

```

Comment:
Why don't you use `std::remove_if` (erase-remove idiom)?

```cpp
const auto it = std::remove_if(_mruPeasants.begin(), _mruPeasants.end(), [&](const auto& p) {
	const auto id = p.PeasantID();
	const auto it = std::find(peasantsToErase.cbegin(), peasantsToErase.cend(), id);
	return it != peasantsToErase.cend();
});
_mruPeasants.erase(it, _mruPeasants.end());
```

Since you repeatedly do a `std::find` over a vector it would be preferable to use a `std::unordered_set` for `peasantsToErase`. Afterwards you can write:
```cpp
_mruPeasants.erase(std::remove_if(_mruPeasants.begin(), _mruPeasants.end(), [&](const auto& p) {
    return peasantsToErase.count(p.PeasantID());
}), _mruPeasants.end());
```

Reply:
That is definitely clearer. For some reason I couldn't find `remove_if` (maybe I was looking on `std::vector` directly).

Reply:
I still don't think there is value in using unordered_set, especially with an expected small number of values to erase, but that is mostly me being stubborn. Really the problem is that the C++ std library is deficient and there is no convenient `.contains` on a vector.

Reply:
You could also view it in a different light: `peasantsToErase` is almost always empty during runtime, so the performance benefit of using a vector normally doesn't apply. Neither container class allocates memory in such a case.
On the other hand if - due to some random misfortune - we do end up having a lot of peasants to erase, the unordered set will perform _asymptotically_ better. It's also more readable in my opinion.

Although I honestly wouldn't block this PR over such a detail.

Reply:
To be more pedantic, `unordered_set.contains` has worst case runtime of linear still :)

Reply:
The hash function we use through the STL (FNV1a) makes that quite unlikely however.
`std::unordered_set` is about ~2x slower than `std::vector` for such worst cases IIRC (with small data sets).

I'm fairly stoked on adding `robin_hood::unordered_map` soon, as it uses a plain vector as the underlying container.

Reply:
switched this and the other cases that used vectors to unordered_sets, and made the _clearOldMruEntries just take a set as parameter.